staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/streaming.go

```diff
@@ -0,0 +1,291 @@
+		if err != nil {
+			return err
+		}
+	}
+	_, err = w.Write([]byte("}\n"))
+	return err
+}
+
+func encodeValue(w io.Writer, prefix []byte, value any, suffix []byte) error {
+	data, err := json.Marshal(value)
+	if err != nil {
+		return err
+	}
+	if len(prefix) != 0 {
+		_, err = w.Write(prefix)
+		if err != nil {
+			return err
+		}
+	}
+	_, err = w.Write(data)
+	if err != nil {
+		return err
+	}
+	if len(suffix) != 0 {
+		_, err = w.Write(suffix)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func streamingEncodeUnstructuredList(w io.Writer, list *unstructured.UnstructuredList) error {
+	_, err := w.Write([]byte(`{`))
+	if err != nil {
+		return err
+	}
+	keys := slices.Sorted(maps.Keys(list.Object))
+	itemsSeen := false
+	for i, key := range keys {
+		value := list.Object[key]
+		if key == "items" {
+			value = list.Items
+			itemsSeen = true
+		}
+		if !itemsSeen && key > "items" {
+			err := encodeKeyValuePair(w, "items", list.Items, true)
+			if err != nil {
+				return err
+			}
+			itemsSeen = true
+		}
+		err := encodeKeyValuePair(w, key, value, i != len(keys)-1 || !itemsSeen)
+		if err != nil {
+			return err
+		}
+	}
+	if !itemsSeen {
+		err := encodeKeyValuePair(w, "items", list.Items, false)
+		if err != nil {
+			return err
+		}
+	}
+	_, err = w.Write([]byte("}\n"))
+	return err
+}
+
+func encodeKeyValuePair(w io.Writer, key string, value any, sep bool) error {
+	switch v := value.(type) {
+	case []runtime.Object:
+		keyData, err := json.Marshal(key)
+		if err != nil {
+			return err
+		}
+		if _, err := w.Write(keyData); err != nil {
+			return err
+		}
+		_, err = w.Write([]byte(":"))
+		if err != nil {
+			return err
+		}
+		if v == nil {
+			_, err := w.Write([]byte("null"))
+			return err
+		}
+		if len(v) == 0 {
+			_, err := w.Write([]byte("[]"))
+			return err
+		}
+		suffix := []byte(",")
+		prefix := []byte("[")
+		for i, item := range v {
+			if i == len(v)-1 {
+				suffix = nil
+			}
+			err := encodeValue(w, prefix, item, suffix)
+			if err != nil {
+				return err
+			}
+			if i == 0 {
+				prefix = nil
+			}
+		}
+		_, err = w.Write([]byte("]"))
+		if sep {
+			_, err = w.Write([]byte(","))
+		}
+		return err
+	case []unstructured.Unstructured:
+		keyData, err := json.Marshal(key)
+		if err != nil {
+			return err
+		}
+		if _, err := w.Write(keyData); err != nil {
+			return err
+		}
+		_, err = w.Write([]byte(":"))
+		if err != nil {
+			return err
+		}
+
+		if len(v) == 0 {
+			_, err := w.Write([]byte("[]"))
+			if err != nil {
+				return err
+			}
+		}
+		suffix := []byte(",")
+		prefix := []byte("[")
+		for i, item := range v {
+			if i == len(v)-1 {
+				suffix = []byte("]")
+			}
+			err := encodeValue(w, prefix, item.Object, suffix)
+			if err != nil {
+				return err
+			}
+			if i == 0 {
+				prefix = nil
+			}
+		}
+		if sep {
+			_, err = w.Write([]byte(","))
+		}
+		return err
+	case string:
+		if v == "" {
+			return nil
+		}
```

Comment:
I would avoid putting special-case omitempty semantics inside this method, and make the caller avoid calling this method if the key/value should be omitted ... not all strings get omitted

Reply:
Or we could implement omitempty for the function.

Reply:
That requires plumbing json tags there or booleans, I'd rather keep this as simple as possible